File: legacy/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SparkfunOdometryLocalizer.java
package org.firstinspires.ftc.teamcode;

import androidx.annotation.NonNull;

import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.localization.Localizer;
import com.qualcomm.hardware.sparkfun.SparkFunOTOS;

public class SparkfunOdometryLocalizer implements Localizer {
    private final SparkFunOTOS odometrySensor;
    private Pose2d _poseEstimate = new Pose2d(0.0, 0.0, 0.0);
    private Pose2d _poseVelocity = new Pose2d();

    public SparkfunOdometryLocalizer(SparkFunOTOS odometrySensor) {
        this.odometrySensor = odometrySensor;
    }

    @NonNull
    @Override
    public Pose2d getPoseEstimate() {
        return _poseEstimate;
    }

    @Override
    public void setPoseEstimate(@NonNull Pose2d value) {
        // No operation
    }

    @Override
    public Pose2d getPoseVelocity() {
        return _poseVelocity;
    }

    @Override
    public void update() {
        _poseEstimate = new Pose2d(odometrySensor.getPosition().x, odometrySensor.getPosition().y, odometrySensor.getPosition().h);
        _poseVelocity = new Pose2d(odometrySensor.getVelocity().x, odometrySensor.getVelocity().y, odometrySensor.getVelocity().h);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SparkfunOdometryLocalizer.kt
package org.firstinspires.ftc.teamcode

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.localization.Localizer
import com.qualcomm.hardware.sparkfun.SparkFunOTOS

class SparkfunOdometryLocalizer (
        private val odometrySensor: SparkFunOTOS
): Localizer {
    private var _poseEstimate = Pose2d(x=0.0, y=0.0, heading=0.0)
    private var _poseVelocity = Pose2d()
    override var poseEstimate: Pose2d
        get() = _poseEstimate
        set(value) {}
    override val poseVelocity: Pose2d?
        get() = _poseVelocity

    override fun update() {
        _poseEstimate = Pose2d(x=odometrySensor.position.x, y=odometrySensor.position.y, heading=odometrySensor.position.h)
        _poseVelocity = Pose2d(x=odometrySensor.velocity.x, y=odometrySensor.velocity.y, heading=odometrySensor.velocity.h)
    }
}